ggml-rpc.cpp

```diff
@@ -0,0 +1,761 @@
+#include "ggml-rpc.h"
+#include "ggml.h"
+#include "ggml-backend-impl.h"
+
+#include <memory>
+#include <string>
+#include <vector>
+#include <unordered_map>
+#include <unordered_set>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <netinet/in.h>
+#include <netinet/tcp.h>
+#include <netdb.h>
+#include <string.h>
+#include <unistd.h>
+
+#define UNUSED GGML_UNUSED
+
+#define GGML_DEBUG 1
+#if (GGML_DEBUG >= 1)
+#define GGML_PRINT_DEBUG(...) printf(__VA_ARGS__)
+#else
+#define GGML_PRINT_DEBUG(...)
+#endif
+
+// RPC data structures
+
+static ggml_guid_t ggml_backend_rpc_guid() {
+    static ggml_guid guid = { 0x00, 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88, 0x99, 0xaa, 0xbb, 0xcc, 0xdd, 0xee, 0xff};
+    return &guid;
+}
+
+struct ggml_backend_rpc_buffer_type_context {
+    int sockfd;
+    std::string name;
+};
+
+struct ggml_backend_rpc_context {
+    std::string endpoint;
+    std::string name;
+    int sockfd;
+    ggml_backend_buffer_type_t buft;
+};
+
+struct ggml_backend_rpc_buffer_context {
+    int sockfd;
+    uint64_t remote_ptr;
+    std::string name;
+};
+
+
+// RPC helper functions
+
+static int socket_connect(const char * host, int port) {
+    struct sockaddr_in addr;
+    int sock = socket(AF_INET, SOCK_STREAM, 0);
```

Comment:
The socket should be closed when returning failure to avoid leaking it.

Reply:
I also need to wrap in a struct and use RAII to properly close it when no longer used